packages/next/src/client/react-client-callbacks/error-boundary-callbacks.ts

```diff
@@ -1,22 +1,41 @@
 // This file is only used in app router due to the specific error state handling.
 
-import type { HydrationOptions } from 'react-dom/client'
+import type { ErrorInfo } from 'react'
 import { getReactStitchedError } from '../components/errors/stitched-error'
 import { handleClientError } from '../components/errors/use-error-handler'
 import { isNextRouterError } from '../components/is-next-router-error'
 import { isBailoutToCSRError } from '../../shared/lib/lazy-dynamic/bailout-to-csr'
 import { reportGlobalError } from './report-global-error'
 import { originConsoleError } from '../components/globals/intercept-console-error'
+import { AppDevOverlayErrorBoundary } from '../components/react-dev-overlay/app/app-dev-overlay-error-boundary'
+import {
+  ErrorBoundaryHandler,
+  GlobalError as DefaultErrorBoundary,
+} from '../components/error-boundary'
+
+export function onCaughtError(
+  err: unknown,
+  errorInfo: ErrorInfo & { errorBoundary?: React.Component }
+) {
+  const errorBoundaryComponent = errorInfo.errorBoundary?.constructor
+
+  const isImplicitErrorBoundary =
+    (process.env.NODE_ENV !== 'production' &&
+      errorBoundaryComponent === AppDevOverlayErrorBoundary) ||
+    (errorBoundaryComponent === ErrorBoundaryHandler &&
```

Comment:
This logic seem easy to break, can we check if the react error lifecyle methods are present then we treat it as an implicit error boundary? `constructor.getDerivedStateFromError || errorBoundary.componentDidCatch`

Reply:
It would actually always be true. `errorInfo.errorBoundary` will always be an error boundary so either cDC or gDSFE must be defined.